### ISSUE TITLE
Can't temporarily disable monitor using external tools

### ISSUE THREAD

**Issue by erahhal** [hyprwm/Hyprland#5978]
### Hyprland Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch  at commit cba1ade848feac44b2eda677503900639581c3f4  ().
Date: 2024-05-04
Tag: , commits: @COMMITS@

flags: (if any)


System Information:
System name: Linux
Node name: upaya
Release: 6.6.30
Version: #1-NixOS SMP PREEMPT_DYNAMIC Thu May  2 14:32:50 UTC 2024


GPU information: 
00:02.0 VGA compatible controller [0300]: Intel Corporation HD Graphics 630 [8086:591b] (rev 04) (prog-if 00 [VGA controller])


os-release: BUG_REPORT_URL="https://github.com/NixOS/nixpkgs/issues"
BUILD_ID="23.11.20240504.651b470"
DOCUMENTATION_URL="https://nixos.org/learn.html"
HOME_URL="https://nixos.org/"
ID=nixos
LOGO="nix-snowflake"
NAME=NixOS
PRETTY_NAME="NixOS 23.11 (Tapir)"
SUPPORT_END="2024-06-30"
SUPPORT_URL="https://nixos.org/community.html"
VERSION="23.11 (Tapir)"
VERSION_CODENAME=tapir
VERSION_ID="23.11"


plugins:
```

</details>

### Bug or Regression?

Bug

### Description

With 0.39.1, you could disable or enable a monitor using other tools, e.g. `kanshi` or `wdisplays`. With 0.40, these tools no longer work. Once a monitor is disabled, hyprland immediately re-enables it. This is a problem for those us that regular get up from their desks and plug and unplug to external monitors multiple times a day. On Discord, it was suggested to just manually update hyprland config, but this isn't a great solution as it involves reloading hyprland, and a rebuild if you are on nixos.

### How to reproduce

Have a laptop plugged into an external monitor. Use `wdisplays` to disable the laptop monitor. Note that it turns back on immediately.

**Comment by vaxerski** [comment]
[patch.txt](https://github.com/hyprwm/Hyprland/files/15274339/patch.txt)
try

**Comment by erahhal** [comment]
Compiled and ran, same problem.

**Comment by LU15W1R7H** [comment]
Can confirm, since hyprland v40, kanshi can no longer disable monitors permenantly. Hyprland immediatly reenables it.

**Comment by neelkarma** [comment]
Same issue here. FWIW, doing a `git bisect` shows that the issue was introduced somewhere between 8a2269272b6a9752cd48097e19c0bc0f3bdab191 and cba1ade848feac44b2eda677503900639581c3f4. (For some reason I couldn't build any of the in-between commits, sorry)

**Comment by innocentDE** [comment]
Does this issue also apply to laptop screens not "hibernating" when closing the lid?

**Comment by tinmarr** [comment]
I use hypridle/hyprlock and that seems to be working fine on my thinkpad.

I also have the other issues faced.

**Comment by levnikmyskin** [comment]
I'm investigating this. 
When the WL client sends the request, `Events::listener_monitorStateRequest` is received. This calls `Renderer::applyMonitorRule`, which then sets `g_pConfigManager->m_bWantsMonitorReload` to `true`.

Now, while `Monitor::onDisconnect` is correctly called, and the monitor is disabled, the issue (unless I missed something) seems to be in the reload itself. 
When `CConfigManager::performMonitorReload` is called, `getMonitorRuleFor` called on the disabled monitor returns the rule where the monitor was still enabled. 
From what I see, the monitor `activeMonitorRule` is changed, but then this has no effect on the rules `CConfigManager` reads?

@vaxerski where would be the most appropriate place to change this so that `CConfigManager`monitor rules are also updated?

**Comment by vaxerski** [comment]
listener_monitorStateRequest is not called here, furthermore, those events are ignored for real displays

**Comment by levnikmyskin** [comment]
Mmm yeah I was testing on a hyprland debug build, running as a window. But the rest of the calls should still happen right?

Anyway, I'll take another look later and come back either to this issue or open a PR if I find a fix :)

**Comment by levnikmyskin** [comment]
Hey everybody, could you check that my PR fixes this problem for you?
https://github.com/hyprwm/Hyprland/pull/6136

**Comment by erahhal** [comment]
Thanks @levnikmyskin, I'm trying to patch 0.40.0 on Nix and it fails to build. I'd try the main branch but that fails to build patched or not. Perhaps someone on a different OS could try it.

**Comment by aacebedo** [comment]
I have tested it on nixos and unfortunately it does not work. External screen keeps being reactivated it when disabled using wdisplays

**Comment by aacebedo** [comment]
> Thanks @levnikmyskin, I'm trying to patch 0.40.0 on Nix and it fails to build. I'd try the main branch but that fails to build patched or not. Perhaps someone on a different OS could try it.

You shall be able to build it with the following overrides in your flake and using master as nixpkgs source:

```
hyprland = prev.hyprland.overrideAttrs (prev: rec {
version = "unstable";
src = final.fetchFromGitHub {
owner = "levnikmyskin";
repo = "hyprland";
fetchSubmodules = true;
rev = "2e74d4e4316fa251e04742ffe2fe2def3a54134b";
hash = "sha256-19tEW0II2ExxBQbfDGg6FL2lAdKzBC4AWADAK4zWyX8=";
};

});

hyprwayland-scanner = prev.hyprwayland-scanner.overrideAttrs
  (prev: rec {
    version = "0.3.8";
    src = final.fetchFromGitHub {
      owner = "hyprwm";
      repo = "hyprwayland-scanner";
      rev = "v0.3.8";
      hash = "sha256-/DwglRvj4XF4ECdNtrCIbthleszAZBwOiXG5A6r0K/c=";
    };

  });

```

**Comment by levnikmyskin** [comment]
@aacebedo that's weird, are you sure you're not using any other tool to manage your monitors? It works for me. I've noticed though that when trying to re-enable the monitor through kanshi (on my branch), that doesn't work anymore, probably due to the changes I've made.

**Comment by aacebedo** [comment]
> @aacebedo that's weird, are you sure you're not using any other tool to manage your monitors? It works for me. I've noticed though that when trying to re-enable the monitor through kanshi (on my branch), that doesn't work anymore, probably due to the changes I've made.

Double checked and confirm that it is not working. wdisplays crashes just after un click apply and the screen gets reenabled right after. I do not have any other software running that would change those settings

**Comment by levnikmyskin** [comment]
Interesting. Could you try with kanshi (that's what I'm using to test)? Just making sure that it is not a tool problem

**Comment by aacebedo** [comment]
> Interesting. Could you try with kanshi (that's what I'm using to test)? Just making sure that it is not a tool problem

Same result with kanshi. Internal laptop screen does not stay off

**Comment by levnikmyskin** [comment]
are you on my latest commit? `f596516f1a9099175bfedbc7f9bf3aab31035bc2`

**Comment by levnikmyskin** [comment]
Yes let's wait for somebody else to also report on this. You could share your logs in the mesntime, could be helpful

**Comment by neelkarma** [comment]
I can confirm that https://github.com/levnikmyskin/Hyprland/commit/07c342bfda933bc89191bef8a63b3d6b6b37da82 has fixed the issue for me - kanshi is now working as expected. Thank you @levnikmyskin!

Here are the logs, if useful: [hyprland.log](https://github.com/hyprwm/Hyprland/files/15376051/hyprland.log)

**Comment by levnikmyskin** [comment]
Happy to hear that :) @aacebedo could you attach your logs? I wouldn't know why it's not working on your side

**Comment by levnikmyskin** [comment]
@neelkarma could you check that it works if you manually ask kanshi to re-enable the monitor? 
For example, in your kanshi config, put something like:
```
profile onlyone {
	output HDMI-something enable ...
	output eDP-1 disable  # this is your internal monitor
}

# probably good to also set positions here
profile both {
	output HDMI-something enable ...
	output eDP-1 enable 
}
```

Then, once your monitor is disabled, run `kanshictl switch both`

**Comment by neelkarma** [comment]
In this logfile, I:

1. Opened hyprland (this starts kanshi as well)
2. Unplugged the HDMI cable (kanshi does not detect this)
3. Open a new terminal and type `kanshi switch undocked` (I have different names for the profiles) - The profile switches successfully
4. Plug the HDMI cable back in (kanshi recognizes this, and applies the correct profile successfully)
5. Exit hyprland

[hyprland1.log](https://github.com/hyprwm/Hyprland/files/15376390/hyprland1.log)

Hope this helps!

Also, I tried on sway, and it worked correctly, so this is probably a hyprland-specific issue.

**Comment by levnikmyskin** [comment]
@neelkarma mmh, I can see from the logs that the HDMI monitor is correctly destroyed by Hyprland. This should indeed send the [finished](https://wayland.app/protocols/wlr-output-management-unstable-v1#zwlr_output_manager_v1:event:finished) event.

Could you run kanshi with `WAYLAND_DEBUG=1 kanshi`, unplug the monitor, and upload the logs here? I'll also check on my machine to see if Hyprland is sending the finished event correctly (looking at the code, it seems like it should)

**Comment by aacebedo** [comment]
Tested on 07c342bfda933bc89191bef8a63b3d6b6b37da82 and confirm it is working with kanshi now.
However, when I unplug HDMI, internal monitor is not reenabled, I managed to do it using wdisplays though.
That is still very unstable and shall be working more seamlessly

**Comment by neelkarma** [comment]
> @neelkarma mmh, I can see from the logs that the HDMI monitor is correctly destroyed by Hyprland. This should indeed send the [finished](https://wayland.app/protocols/wlr-output-management-unstable-v1#zwlr_output_manager_v1:event:finished) event.
> 
> Could you run kanshi with `WAYLAND_DEBUG=1 kanshi`, unplug the monitor, and upload the logs here? I'll also check on my machine to see if Hyprland is sending the finished event correctly (looking at the code, it seems like it should)

[kanshi.log](https://github.com/hyprwm/Hyprland/files/15377250/kanshi.log)

**Comment by levnikmyskin** [comment]
Could you guys try now (commit [`2ad0038` (#6136)](https://github.com/hyprwm/Hyprland/pull/6136/commits/2ad003810abacb30fa943aaac5ff793f36562f2a))? I think the problem was that Hyprland was not sending the [done](https://wayland.app/protocols/wlr-output-management-unstable-v1#zwlr_output_manager_v1:event:done) event after the `finished` event. So, kanshi was basically waiting forever to receive it.

**Comment by Aleksanaa** [comment]
> Edit: Interestingly, `kanshi` doesn't seem to be working fully - It successfully applies the initial config, but it doesn't enable the laptop display upon unplugging my monitor. Looking at `kanshi`'s output, it seems that it isn't detecting the HDMI cable being unplugged. Not sure if this is related to this issue or something entirely different. Will test on swaywm to see if this is a kanshi-specific or Hyprland-specific issue.

That seems like another issue that has existed much longer than this issue. In fact, I've never got hyprland working in this way. I think we should probably test with `wlr-randr --output <output> off` (and back on) at this time instead of testing whether recovering monitor is available during detaching output.

**Comment by aacebedo** [comment]
> Could you guys try now (commit [`2ad0038` (#6136)](https://github.com/hyprwm/Hyprland/pull/6136/commits/2ad003810abacb30fa943aaac5ff793f36562f2a))? I think the problem was that Hyprland was not sending the [done](https://wayland.app/protocols/wlr-output-management-unstable-v1#zwlr_output_manager_v1:event:done) event after the `finished` event. So, kanshi was basically waiting forever to receive it.
> 
> EDIT: PLEASE if it doesn't work, attach hyprland logs + kanshi logs (with `WAYLAND_DEBUG=1` as before)

Seems to be ok with one caveat: If you start the computer with the HDMI attached it fails to turn off the internal screen. But it works if after you have started hyprland, you unplug and replug the HDMI.

**Comment by Aleksanaa** [comment]
> Could you guys try now (commit [`2ad0038` (#6136)](https://github.com/hyprwm/Hyprland/pull/6136/commits/2ad003810abacb30fa943aaac5ff793f36562f2a))? I think the problem was that Hyprland was not sending the [done](https://wayland.app/protocols/wlr-output-management-unstable-v1#zwlr_output_manager_v1:event:done) event after the `finished` event. So, kanshi was basically waiting forever to receive it.
> 
> EDIT: PLEASE if it doesn't work, attach hyprland logs + kanshi logs (with `WAYLAND_DEBUG=1` as before)

Oh wait, you fixed it! YOU FIXED IT!

**Comment by Aleksanaa** [comment]
Now when I unplug the HDMI cable, the built-in display automatically activates. Also the issue described is fixed.

**Comment by levnikmyskin** [comment]
@Aleksanaa good to hear it's working for you. Could you try the situation described by @aacebedo. I can't reproduce it, on my machine it works even if HDMI is plugged in before logging into hyprland

**Comment by Aleksanaa** [comment]
For me it works. I mean, if you plug in the HDMI cable before starting hyprland, the built-in display will be automatically turned off after kanshi starts.

**Comment by aacebedo** [comment]
I'll try producing some logs

**Comment by aacebedo** [comment]
here it is:
[kanshi.log](https://github.com/hyprwm/Hyprland/files/15379161/kanshi.log)

In this log, I kept the screen attached and started hyprland. Then I started kanshi, It applies home-office profile which shall turn off the internal screen but does not. Then I unplug the screen and replug it, this time the internal screen turns off.

**Comment by levnikmyskin** [comment]
Can't see much from these logs unfortunately. I'd need the hyprland logs too. Also, probably running kanshi with:
`exec-once = sleep 1 && kanshi`
might be a good workaround to this

**Comment by neelkarma** [comment]
The latest commit (https://github.com/levnikmyskin/Hyprland/commit/2ad003810abacb30fa943aaac5ff793f36562f2a) is causing a Hyprland crash for me.

Here's what I did:

1. Start hyprland (without starting kanshi)
2. Open a new terminal and start kanshi
3. Unplug HDMI cable
4. Observe crash

Here are the logs:
[hyprland.log](https://github.com/hyprwm/Hyprland/files/15385446/hyprland.log)
[kanshi.log](https://github.com/hyprwm/Hyprland/files/15385447/kanshi.log)

Considering everyone else seems to have no issues, this might be a completely different problem, but I thought I'd check here just in case.

**Comment by levnikmyskin** [comment]
@neelkarma from your logs there's nothing useful unfortunately. It seems something is crashing when returning from the unsafe state. This in my opinion is _not_ related to this issue: kanshi is working correctly, and communication between hyprland and the wayland client seems correct.

You will need to open another issue. Valuable information for the crash will be:
	1. whether you're using any plugin (can't see it from your logs);
	2. the coredump (see [here](https://github.com/hyprwm/Hyprland/blob/main/docs/ISSUE_GUIDELINES.md#obtaining-the-debug-hyprland-coredump)).

For everybody else, I'll wait until thursday to ping Vaxry to start the merging process. Please, if you're still experiencing issues related to this, let me know

**Comment by erahhal** [comment]
I can also confirm that the latest patch is working for me.

**Comment by leiserfg** [comment]
@fufexan this is sadly broken in nixos 24-05, do you think it can be backported?

**Comment by fufexan** [comment]
I've created some confusion with the previous backport (which downgraded the version due to a regression). Not even sure if all the regressions are fixed currently. I'll update and backport when the next version is out.

**Comment by sjdrc** [comment]
Scratch that. I stopped using the flake and updated nixpkgs and now everything works as expected. Thought I had updated all my inputs but I guess I had something messed up.
